spec/controllers/user_spec.rb

```diff
@@ -0,0 +1,89 @@
+
+RSpec.describe Sinatra::App:: SignUp do
+  include Rack::Test::Methods
+  include Warden::Test::Helpers
+  let(:user) { build :user }
+  let(:admin)  { build :admin }
+  def app
+    @app = App
+  end
+
+  context 'with valid credentials' do
+    it 'register a user, and send email' do
+      post '/signup', params = { user:
```

Comment:
Since you have already defined a _factory_ for users, you can use FactoryBot's `attributes_for` method rather than having to repeat these params in all your tests.

```ruby
 post '/signup', params = { user: attributes_for(:user) }
```